### ISSUE TITLE
Prisma 2 on Server?

### ISSUE THREAD

**Issue by mishushakov** [prisma/prisma2#100]
Hi guys,
i searched for a GraphQL layer for my SQL database and prisma seemed like the best deal out there, with lots of documentation, huge community and experienced developers.

Some things I liked about Prisma:
- SDL schema (+relationships)
- Convenient Admin Dashboard (is there a way to disable it?)
- Convenient GraphQL UI (is there a way to disable it?)
- Convenient CLI
- Server-side security with JWT tokens
- Ability to have multiple "services" within same database (which is amazing for micro-services!)
- Generated type-safe clients (why would we need them, when we've got GraphQL!?!)
- GraphQL, GraphQL, GraphQL!!!

I then browsed your website and discovered, that you're working on a second version of Prisma and are looking to move some server components to users' machines: 

![hWOqmYA](https://user-images.githubusercontent.com/10400064/60734165-4e83bd00-9f4f-11e9-8928-90b000dd50cd.png)

Does that mean, we will have to bundle the Prisma executable with our app and build GraphQL server ourselves? Why would we need Prisma then, if we could do it before with any ORM/ODM?

If it's true, than it looks like a major step back, because the primary value of Prisma (for me at least) is to have a software, that turns a 20-yrs old SQL database into a modern, real-time, GraphQL database.

I don't want to write Server-Side side code since Firebase. Now, I want to have Prisma to serve GraphQL API for my RDBMS, so I could focus on writing my Web/Mobile Applications.

Thank you guys! Please don't get rid of the Server-side GraphQL version :D

**Comment by auderer** [comment]
No, its not being moved to users' machines! You still have your own backend server which handles GraphQL. The difference in Prisma2 is that you no longer need the *additional* Prisma server to handle database access.

In Prisma1, you needed TWO servers - one for your own backend implementation, and one for Prisma's query engine.

The query engine is not being moved onto the front-end, but rather bundled onto your existing server.

This graph may demonstrate that a bit better:

Prisma2:

![image](https://user-images.githubusercontent.com/28285686/60774013-a2bba800-a0db-11e9-86c9-bbefbbd46b1d.png)

Compared to the old Prisma 1:

![image](https://user-images.githubusercontent.com/28285686/60774021-b961ff00-a0db-11e9-8c28-f86799958978.png)

**Comment by mishushakov** [comment]
Hey @auderer thank you for your comment!

Yes, I'm getting the concept. I just want to address certain type of problems, that may affect some customers (like me) if you do it this way.

What you're telling me, that I need to have a backend and Photon executable that will talk to the Database. That creates a problem:

1) I don't want to have/code/maintain/serve any Backends
2) My Application runs on Browsers/Mobile/IoT/Edge, where resources are limited and therefor I can't physically bundle a backend with the distribution
3) It's much more cost-efficient for me to run a single Prisma Instance, that serves as many services I want, rather than install each per service.
4) I don't want to rely on the auto-generated client, I just need a GraphQL API for my database, which other clients, that you don't yet support (like Swift, Python ...) can talk to

Here is how I use Prisma:

![Artboard](https://user-images.githubusercontent.com/10400064/60798397-5cb12380-a171-11e9-8bf4-3ab1e5627070.png)

As you see, there is nothing in between the Client and Prisma Server, this way Prisma is effectively a **serverless backend**. Don't you think there is a market for it?

What makes Prisma so unique, is that it is really a *GraphQL database*, that happens to run on top of existing Database-solutions.

I'd actually, really love to see Prisma becoming a Database on it's own (and I'd like to help) someday. This could be the next Firebase.

**Comment by mishushakov** [comment]
It now doesn't make any sense to have Prisma, because Photon's Rust-executable will talk to JS client via a protocol, which is likely not GraphQL and if I want to have GraphQL API, I will have to code it for my own... What is bullshit (sorry). We don't need *yet another ORM*. We need **no ORM**!!!

**Comment by Torsten85** [comment]
I'm currently facing the same situation.

I'm researching a Postgres abstract for a new project and came across prisma which had some great features I always wanted (GraphQL!). But the development of prisma2 is going in a wrong direction for me. 

As I see it, I have to implement graphql myself from scratch... I already did this with `node-postgres`, so why do I need prisma2? 

Having a standalone prisma1 service is no big deal in our k8s cluster (and we can scale it independently from our other backend!)

Will prisma2 replace prisma1 or will both be supported in the future?

**Comment by pantharshit00** [comment]
Hi all,

Prisma is not a Backend as a Service or a GraphQL API generator. Prisma is a simply an ORM replacement which will give you tools to access your database. Then you can make any kind of APIs using the data retrieved by Prisma. That might be a GraphQL API, gRPC API or a REST API.


If you want a GraphQL API generator or a BaaS you can look into tools like Hasura or AppSync. Prisma is meant to achieve a different goal. You can also use Prisma in conjunction with something like Hasura to implement your custom database logic.


We have written a blog post on this to clarify this confusion: https://www.prisma.io/blog/prisma-and-graphql-mfl5y2r7t49c

**Comment by soullivaneuh** [comment]
> Many of the tools we've built have become the default in many GraphQL development workflows, such as the GraphQL Playground, graphql-yoga and GraphQL Nexus (built by Tim Griesser).

In a nutshell, the Prisma1 built-in GraphQL server moved onto several projects.

Am I right @pantharshit00?

Still wondering, why targeting 3 tools? What are the difference?

In anyway, your tools still look wonderful, thanks for that!

**Comment by soullivaneuh** [comment]
I also have a suggestion: Why not building a demo website with our Prisma2 ecosystem?

It would certainly help us to understand what you are trying to change. :+1:

**Comment by balthild** [comment]
IIRC, the GraphQL API in Prisma 1 is just an internal protocol that Prisma clients communicate with the database, and it's not intended to be called by application's backend server. That's really unexpected for me.

For those backends built on top of languages which has not supported by Prisma Client yet (for example, Rust), programmers must malually call the GraphQL API -- the language-agnostic way -- in order to gain type safety when interacting with database. Since it's impossible to ask developers of Prisma 2 to support all programming languages, the removal of GraphQL API will block such projects migrating to Prisma 2. Will there be recommended  methods to generate GraphQL API which is compatiable with that used by Prisma 1?

**Comment by nikolasburk** [comment]
Hey @balthild, thanks a lot for your comment!

> IIRC, the GraphQL API in Prisma 1 is just an internal protocol that Prisma clients communicate with the database, and it's not intended to be called by application's backend server. 

That's correct 👍 the GraphQL API is an implementation detail that developers shouldn't worry about. They should only be concerned with the Prisma client libraries that provide a way to access their databases from a programming language. 

> For those backends built on top of languages which has not supported by Prisma Client yet (for example: Rust, C#, Kotlin, etc...), programmers must malually call the GraphQL API -- the language-agnostic way -- in order to gain type safety when interacting with database.

We indeed want to support as many programming languages as possible with Prisma Client. Right now, Prisma Client 2.0 only works for JavaScript and TypeScript, but we're already working on the Golang version of it as well. With Prisma 1, it was certainly possible to use Prisma's GraphQL API from applications that were built in other languages  – although I'm not aware of many people taking this approach, would love to hear from you whether you've seen this be a major use case? If we get the feedback that this is a feature that's strongly requested by our community we might consider adjusting our direction there, right now our focus right now is on the Node.js ecosystem. 

Also note that there still is a language-agnostic query engine in Prisma 2 (written in Rust) that receives requests from Prisma Client. So, theoretically if you wanted to use Prisma 2 with another programming language, you can still do that by communicating with that query engine. This is not as trivial as calling a GraphQL API as in Prisma 1 though, since the query engine is now running as a _sidecar_ process on your application server and not as a standalone infrastructure component (as was the case in Prisma 1 with the Prisma server).

Does that help? 🙂

**Comment by balthild** [comment]
Hello @nikolasburk . Thank you for your reply. It's helpful.

As you say, I can still talk with the query engine by myself in Prisma 2. Although it's not as easy as GraphQL, it will stll be way to migrate to Prisma 2, whose performance is better.

But it seems that the query engine does not provide an API doc currently. It would be better if you could describe how a Prisma client communicates with the engine, and it will make working without official Prisma Clients easier, whether developing an application in Rust, or creating a third-party Prisma Client.

**Comment by jylin** [comment]
Would it be possible to create a new tool/product that generated a standard GraphQL server based on the schema? I think this is part of the functionality that was lost with the new Prisma 2 design: it's not longer possible to create a browser-client only app like this: https://www.prisma.io/tutorials/build-react-graphql-app-with-fetch-ct19/

**Comment by balthild** [comment]
@jylin I think [nexus-prisma](https://github.com/prisma-labs/nexus-prisma) is suitable for this scenario.

**Comment by divideby0** [comment]
For anyone visiting this issue more recently, @AhmedElwya's Pal.js framework does a lot of the stuff you might be looking for to bridge the gap between Prisma 2 and the Prisma 1 GraphQL API:

https://paljs.com/

It leverages Nexus but provides Rails-style generators to create all your GQL interfaces from its CLI. For example, if you had a Book and Author models in Prisma, running `pal g Book,Author crud` would create all the queries, mutations, etc for you to wire into your ApolloServer schema.

We've been working with Paljs (and Ahmed) quite a bit lately and have found it to save us a lot of time.

**Comment by johnlister** [comment]
For me it would be good to run the Prisma 2 as an actual sidecar container in docker/kubernetes such that it can be managed separately. Also as mentioned above, from a performance point of view, when running potentially hundreds of microservices, it makes sense to me to run a cluster of the Prisma 2 "server" and point all of my microservices at the cluster. This would potentially also reduce the number of active pooled connections  open against the DB, but still have all the benefits of splitting the DB engine from the client code
